lectures/sympy.md

```diff
@@ -0,0 +1,861 @@
+$$
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 17.5%
+---
+# Define the present value equations
+PV_highschool = w/R + w*(1 + g)/R**2 + w*(1 + g)**2/R**3 + w*(1 + g)**3/R**4
+PV_college = D + phi*w*(1 + g)**4/R + phi*w*(1 + g)**5 / \
+    R**2 + phi*w*(1 + g)**6/R**3 + phi*w*(1 + g)**7/R**4
+```
+
++++ {"user_expressions": []}
+
+### Defining the Indifference Equation
+The indifference equation represents the condition that a worker is indifferent between going to college or not. This is given by the equation $PV_h$ = $PV_c$
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 17.5%
+---
+# Define the indifference equation
+indifference_eq = Eq(PV_highschool, PV_college)
+```
+
++++ {"user_expressions": []}
+
+We can now solve the indifference equation for the wage gap $\phi$
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 65%
+---
+# Solve for phi
+solution = solve(indifference_eq, phi)
+
+# Simplify the solution
+solution = simplify(solution[0])
+solution
+```
+
++++ {"user_expressions": []}
+
+To compute a numerical value for $\phi$, we replace symbols $w$, $R$, $g$, and $D$ with specific numbers. 
+
+Here we use $w = 1$, $R = 1.05$, $g = 0.02$, and $D = 0.5$. 
+
+Substituting them into the equation, we find a specific value for $\phi$
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 20%
+---
+# Substitute specific values
+solution_num = solution.subs({w: 1, R: 1.05, g: 0.02, D: 0.5})
+solution_num
+```
+
++++ {"user_expressions": []}
+
+The wage of a college graduate is approximately 0.797 times the wage of a high school graduate.
+
++++ {"user_expressions": []}
+
+```{exercise}
+:label: sympy_ex1
+
+L'Hôpital's rule states that for two functions $f(x)$ and $g(x)$, if $\lim_{x \to a} f(x) = \lim_{x \to a} g(x) = 0$ or $\pm \infty$, then
+
+$$
+\lim_{x \to a} \frac{f(x)}{g(x)} = \lim_{x \to a} \frac{f'(x)}{g'(x)}
+$$
+
+Use SymPy to verify L'Hôpital's rule for the following functions
+
+$$
+f(x) = \frac{y^x - 1}{x}
+$$
+
+as $x$ approaches to $0$
+```
+
+```{solution-start} sympy_ex1
+:class: dropdown
+```
+
++++
+
+Let's define the function first
+
+```{code-cell} ipython3
+f_upper = y**x - 1
+f_lower = x
+f = f_upper/f_lower
+f
+```
+
+Sympy is smart enough to solve this limit
+
+```{code-cell} ipython3
+lim = limit(f, x, 0)
+lim
+```
+
+We compare the result suggested by L'Hôpital's rule
+
+```{code-cell} ipython3
+lim = limit(diff(f_upper, x)/
+            diff(f_lower, x), x, 0)
+lim
+```
+
+```{solution-end}
+```
+
+```{exercise}
+:label: sympy_ex2
+
+Maximum likelihood estimation (MLE) is a method to estimate the parameters of a statistical model. 
+
+It usually involves maximizing a log-likelihood function and solving the first-order derivative.
+
+In this exercise, we use SymPy to compute the MLE of a binomial distribution.
+```
+
+```{solution-start} sympy_ex2
+:class: dropdown
+```
+
+First, we define the binomial distribution
```

Comment:
Sympy provides this too: https://docs.sympy.org/latest/modules/stats.html#sympy.stats.Binomial

Reply:
Many thanks @Smit-create for your review,

I acknowledged that SymPy has a module on statistics, but I did not include it because the lecture gets super long now : ). 

Maybe I should include a sentence or two to run through a quick example in conjunction with the implementations from scratch.